simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java

```diff
@@ -0,0 +1,235 @@
+                }
+            }
+            int q = (int) (Math.round(event.missPenalty() / Math.pow(nQ, w)) + CH[w]) % nQ;
+            node.updateLoc(w, q);
+            wheels[w].add(q, node.event);
+        }
+
+    }
+
+    private void migration(int idx) {
+        CH[idx] = (CH[idx] + 1) % nQ;
+        if (CH[idx] == 0 && idx + 1 < nW) {
+            migration(idx + 1);
+        }
+        while (wheels[idx].queueNotEmpty(CH[idx])) {
+            AccessEvent p = wheels[idx].evict(CH[idx]);
+            int cr = ((int) p.missPenalty()) % ((int) Math.pow(nQ, idx));
+            int q = (int) (Math.round(cr / Math.pow(nQ, idx - 1)) + CH[idx - 1]) % nQ;
+            wheels[idx - 1].add(q, p);
+            Node node = data.get(p.key());
+            node.updateLoc(idx - 1, q);
+        }
+    }
+
+    @Override
+    public PolicyStats stats() {
+        return policyStats;
+    }
+
+    static class GDWheelSettings extends BasicSettings {
+        public GDWheelSettings(Config config) {
+            super(config);
+        }
+
+        public int numberOfWheels() {
+            return config().getInt("gd-wheel.nw");
+        }
+
+        public int numberOfQueues() {
+            return config().getInt("gd-wheel.nq");
+        }
+    }
+
+    static class CostWheel {
+        List<PriorityQueue<AccessEvent>> wheel;
+        int current_index;
+        int nQ;
+
+        public CostWheel(int nQ) {
+            this.wheel = new ArrayList<>();
+            Comparator<AccessEvent> comparator = new EventComparator();
+            for (int i = 0; i < nQ; i++) {
+                this.wheel.add(new PriorityQueue<>(comparator));
+            }
+            this.current_index = 0;
+            this.nQ = nQ;
+        }
+
+        public int getNextIndex() {
+            for (int i = current_index; i < current_index + nQ; i++) {
+                int j = i % nQ;
+                if (!wheel.get(j).isEmpty()) {
+                    current_index = j;
+                    break;
+                }
+            }
+            return current_index;
+        }
+
+        public AccessEvent evict(int i) {
+            PriorityQueue<AccessEvent> queue = wheel.get(i);
+            return Objects.requireNonNull(queue.poll());
+        }
+
+        public void remove(int i, AccessEvent e) {
+            PriorityQueue<AccessEvent> queue = wheel.get(i);
+            queue.remove(e);
+        }
+
+        public void add(int i, AccessEvent e) {
+            PriorityQueue<AccessEvent> queue = wheel.get(i);
+            queue.add(e);
+        }
+
+        public boolean queueNotEmpty(int i) {
+            PriorityQueue<AccessEvent> queue = wheel.get(i);
+            return !queue.isEmpty();
+        }
+    }
+
+    static class EventComparator implements Comparator<AccessEvent> {
+        @Override
+        public int compare(AccessEvent e1, AccessEvent e2) {
+            return Double.compare(e1.missPenalty(),e2.missPenalty());
+        }
+    }
```

Comment:
fyi, Java 8 introduced a one-liner for these cases by using lambdas, 
`Comparator.comparingDouble(AccessEvent:: missPenalty)`

Reply:
Cool, I like this syntax better, I'll adopt it.